File: leetcode/932.beautiful-array.cpp
//  Tag: Array, Math, Divide and Conquer
//  Time: O(N)
//  Space: O(N)
//  Ref: -
//  Note: dp

//  An array nums of length n is beautiful if:
//  
//  nums is a permutation of the integers in the range [1, n].
//  For every 0 <= i < j < n, there is no index k with i < k < j where 2 * nums[k] == nums[i] + nums[j].
//  
//  Given the integer n, return any beautiful array nums of length n. There will be at least one valid answer for the given n.
//   
//  Example 1:
//  Input: n = 4
//  Output: [2,1,4,3]
//  Example 2:
//  Input: n = 5
//  Output: [3,1,2,5,4]
//  
//   
//  Constraints:
//  
//  1 <= n <= 1000
//  
//  

class Solution {
public:
    unordered_map<int, vector<int>> cache;
    vector<int> beautifulArray(int n) {
        if (n == 1) {
            return vector<int>{1};
        }

        if (cache.count(n)) {
            return cache[n];
        }

        vector<int> even = beautifulArray(n / 2);
        vector<int> odd = beautifulArray(n - n / 2);

        vector<int> res;
        for (auto k: even) {
            res.push_back(2 * k);
        }

        for (auto k: odd) {
            res.push_back(2 * k - 1);
        }

        cache[n] = res;
        return res;
    }
};

class Solution {
public:
    vector<int> beautifulArray(int n) {
        vector<vector<int>> dp(n + 1);
        dp[1] = {1};

        for (int i = 2; i <= n; i++) {
            int even = i / 2;
            int odd = i - i / 2;

            for (auto k : dp[even]) {
                dp[i].push_back(2 * k);
            }

            for (auto k : dp[odd]) {
                dp[i].push_back(2 * k - 1);
            }
        }

        return dp[n];
    }
};

class Solution {
public:
    vector<int> beautifulArray(int n) {
        vector<int> res = {1};

        while (res.size() < n) {
            vector<int> tmp;
            for (auto k: res) {
                if (2 * k <= n) {
                    tmp.push_back(2 * k);
                }
            }

            for (auto k: res) {
                if (2 * k - 1 <= n) {
                    tmp.push_back(2 * k - 1);
                }
            }

            res = tmp;
        }

        return res;
    }
};

File: leetcode/932.beautiful-array.py
#  Tag: Array, Math, Divide and Conquer
#  Time: O(N)
#  Space: O(N)
#  Ref: -
#  Note: dp

#  An array nums of length n is beautiful if:
#  
#  nums is a permutation of the integers in the range [1, n].
#  For every 0 <= i < j < n, there is no index k with i < k < j where 2 * nums[k] == nums[i] + nums[j].
#  
#  Given the integer n, return any beautiful array nums of length n. There will be at least one valid answer for the given n.
#   
#  Example 1:
#  Input: n = 4
#  Output: [2,1,4,3]
#  Example 2:
#  Input: n = 5
#  Output: [3,1,2,5,4]
#  
#   
#  Constraints:
#  
#  1 <= n <= 1000
#  
#  

from functools import lru_cache
class Solution:
    @lru_cache(maxsize=None)
    def beautifulArray(self, n: int) -> List[int]:

        if n == 1:
            return [1]

        even = self.beautifulArray(n // 2)
        odd = self.beautifulArray(n - (n // 2))

        res = []
        for k in even:
            res.append(2 * k)

        for k in odd:
            res.append(2 * k - 1)

        return res
    
class Solution:
    def beautifulArray(self, n: int) -> List[int]:

        dp = {1 : [1]}
        for i in range(2, n + 1):
            even = i // 2
            odd = i - even

            dp[i] = []
            for k in dp[even]:
                dp[i].append(2 * k)

            for k in dp[odd]:
                dp[i].append(2 * k - 1)

        return dp[n]
    
class Solution:
    def beautifulArray(self, n: int) -> List[int]:

        res = [1]
        while len(res) < n:
            even = [2 * i for i in res if 2 * i <= n]
            odd = [2 * i - 1 for i in res if 2 * i - 1 <= n]
            res = even + odd

        return res